controller/lib/hal/stepper.cpp

```diff
@@ -0,0 +1,820 @@
+  dma->channel[C4].config.dir = DmaChannelDir::MEM_TO_PERIPHERAL;
+  dma->channel[C4].config.circular = 0;
+  dma->channel[C4].config.perInc = 0;
+  dma->channel[C4].config.memInc = 1;
+  dma->channel[C4].config.psize = 0;
+  dma->channel[C4].config.msize = 0;
+  dma->channel[C4].config.priority = 0;
+  dma->channel[C4].config.mem2mem = 0;
+
+  Hal.EnableInterrupt(InterruptVector::DMA2_CH3, IntPriority::STANDARD);
+
+  StepMotor::OneTimeInit();
+}
+
+// Do some basic init of the stepper motor chips so we can
+// make them spin the motors
+void StepMotor::OneTimeInit() {
+  uint32_t val;
+  for (int i = 0; i < total_motors_; i++) {
+
+    StepMotor *mtr = StepMotor::GetStepper(i);
+
+    // Reset the chip to default values
+    mtr->Reset();
+
+    // Get the first gate config register of the powerSTEP01.
+    // This is actually the config register on the L6470
+    mtr->GetParam(StepMtrParam::GATE_CFG1, &val);
+
+    // If this is at the default config register value for
+    // the L6470 then I don't need to do any more configuration
+    if (val == 0x2E88)
+      continue;
+
+    // Configure the two gate config registers to reasonable values
+    //
+    // GATE_CFG1 xxxxxxxxxxxxxxxx
+    //           ...........\\\\\ - time at constant current 3750ns (0x1D)
+    //           ........\\\------- Gate current 96mA (7)
+    //           .....\\\---------- Turn off boost time 1uS (7)
+    //           ....\------------- Watch dog enable (1)
+    //           \\\\-------------- reserved
+    mtr->SetParam(StepMtrParam::GATE_CFG1, 0x0FFD);
+
+    // GATE_CFG2 xxxxxxxx
+    //           ...\\\\\---------- Dead time 1000ns (7)
+    //           \\\--------------- Blanking time 1000ns (7)
+    mtr->SetParam(StepMtrParam::GATE_CFG2, 0xF7);
+  }
+}
+
+// Convert a velocity from Deg/sec units to the value to program
+// into one of the stepper controller registers
+float StepMotor::DpsToVelReg(float vel, float cnv) {
+
+  // Convert to steps / sec
+  float step_per_sec = vel * static_cast<float>(steps_per_rev_) / 360.0f;
+
+  // Multiply that by the register specific conversion factor
+  return cnv * step_per_sec;
+}
+
+// Convert a velocity from a register value to deg/sec
+float StepMotor::RegVelToDps(int32_t val, float cnv) {
+  return static_cast<float>(val) * 360.0f /
+         (cnv * static_cast<float>(steps_per_rev_));
+}
+
+// Read the current absolute motor velocity and return it
+// in deg/sec units
+// Note that this value is always positive
+StepMtrErr StepMotor::GetCurrentSpeed(float *ret) {
+  uint32_t val;
+  StepMtrErr err = GetParam(StepMtrParam::SPEED, &val);
+  *ret = RegVelToDps(val, cvt_vel_crnt_speed_reg);
+  return err;
+}
+
+// Set the motor's max speed setting in deg/sec
+//
+// NOTE - The motor must be disabled to set this
+StepMtrErr StepMotor::SetMaxSpeed(float dps) {
+  if (dps < 0)
+    return StepMtrErr::BAD_VALUE;
+
+  uint32_t speed =
+      static_cast<uint32_t>(DpsToVelReg(dps, cvt_vel_max_speed_reg));
```

Comment:
I wonder if Angle (degrees) and AnglePerSec should be units?

Reply:
I agree that would be nice.
Could that be an improvement made in another PR?  This PR is already pretty big but I've managed to contain it to just the HAL library.  I'm reluctant to start spilling over into other libraries.
Making motor speed and position units more generic would be nice, but could be handled in it's own smaller PR

Reply:
Sure.